### ISSUE TITLE
Cache function calls by default

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#7428]
## Problem

Dagger has great caching, but Dagger Functions don't fully benefit from it, because their runtime containers are not cached.

This has several consequences:

1. Functions that perform compute-intensive work themselves, via native code, are most affected: they are re-run each time, adding a major performance penalty.
2. Functions that are not compute-intensive (the majority) are less affected, but still pay a small performance tax at each execution, because of the overhead of executing the runtime container each time. For complex pipelines this can add up.
3. Module initialization is also slower, because SDKs themselves are modules and calling the initialization functions incurs the same performance tax.

The result is that most `dagger call` commands are noticeably slower than they would be, because of insufficient caching.

## Solution

Cache function calls by default!

That would be trivial to do, there's just one problem...

### Cache invalidation

Sometimes we want a function call to *not* be cached, for a variety of reasons:

- Perhaps the function is not "pure" and will not work properly if cached
- Perhaps the function *can* be cached, but the operator wants to control when to cache it and for how long]
- Perhaps the function requires custom caching behavior - for example a dependency lookup might be useful to persist in a lockfile accessible to the user

In order to cache function calls by default, we need a minimal design to allow disabling that caching. That design should 1) be as simple as possible so we can ship fast - but 2) be future-proof so that we don't need to break it later.

May the bikeshedding begin :)

**Comment by shykes** [comment]
FYI @sipsma @vito here is a penciled in proposal based on the discord discussion. I kept the word "pure" out of it, because I think it might more loaded than we originally thought (ie. from a functional programming purist's perspective, none of these cacheable functions qualify as "pure". Perhaps best not to kick the hornet nest...)

**Comment by sipsma** [comment]
In the interest of being able to ship something quickly and then iterate, I think it's worth considering starting with TTL settings as being defined as properties of the function being invoked (i.e. by the author, rather than the caller).

Starting there is relatively straightforward implementation-wise. You could do it without having to go all the way into lock files or anything like that; could just re-use existing cache mechanisms (just need to scope cache to a time interval in the same way you already can do that in "user-space", just automatically now).

Supporting the caller setting the TTL opens up a pandora's box:
1. For the caching to work as expected for each caller (who may now all call the same function with different TTLs), I think we'd need to support per-client lock files (and some way of specifying which one to use)
   * on top of the previously discussed trickiness of lock files like pruning
2. Figure out how to implement cases like "multiple calls in the same session that call the same function with different TTLs"
   * To be clear this is tricky but certainly solvable.

Starting out with the TTL being a static property of a function would avoid all that and thus be way quicker to get going.

To be clear, I'm not discounting the value of client-side settings for this, this is more just about what order we stage the features out.

**Comment by shykes** [comment]
I am happy to cut the TTL feature altogether from the initial implementation. I prefer that to letting the caller define its own TTL.

**Comment by shykes** [comment]
I'm also proposing to leave lock files out of this proposal as well. They would be added later as part of "lookup functions" or some other expansion of the design.

**Comment by sipsma** [comment]
Yeah starting with no TTL is even simpler, so I'm good with starting there too. 

> I am happy to cut the TTL feature altogether from the initial implementation. I prefer that to letting the caller define its own TTL.

Did you mean you prefer that to letting the *callee* (as in the module author) define its own TTL? If so I'd disagree since putting the onus *always* on the caller is going to lead to a lot of repetitive code (a given version of a function often gets called a lot of places, whereas its only authored once) and a lot of situations in which the caller doesn't know what a good TTL is, but the author does. I think a model where authors can define a default TTL if they want and then callers can override it if they want would be the most usable.

**Comment by shykes** [comment]
> Yeah starting with no TTL is even simpler, so I'm good with starting there too.

🚀 

> > I am happy to cut the TTL feature altogether from the initial implementation. I prefer that to letting the caller define its own TTL.
> 
> Did you mean you prefer that to letting the _callee_ (as in the module author) define its own TTL?

Yes, my bad - I corrected the typo.

> If so I'd disagree since putting the onus _always_ on the caller is going to lead to a lot of repetitive code (a given version of a function often gets called a lot of places, whereas its only authored once) and a lot of situations in which the caller doesn't know what a good TTL is, but the author does.

Well, the person who knows best what TTL to set, and will most often feel the need to control it, is the end-user. So the most important affordance for setting TTL is actually in the CLI: `dagger call --ttl` seems like the simplest implementation.

From there, allowing a caller to configure TTL in code (the part you object to) is not really a must-have, but felt like a natural continuity, because the pattern already exists in the wild, with exec cache-busters. So this would be a more general, cleaner version of the cache-buster pattern.

When we discussed it with @sagikazarmark and @vito , we couldn't find good examples of a function developer knowing in advance what TTL to set for their function.

Since we don't yet have consensus on how to handle TTLs, all the more reason to cut it from the first phase of implementation :) I will update the proposal to clarify that.

**Comment by shykes** [comment]
> I'm retrieving some remote resource (downloading a file, whatever), but doing so is an expensive operation and/or want to try to avoid rate limiting. The module author knows the most about the remote resource, so they would be best fit to decide the default TTL in this case. Callers can override if they want but if they are not familiar with the inner details of what they are retrieving (i.e. they are on a different team and just calling another team's module), then they don't have to care.
> 
> Basically, one of the fundamental benefits of modules is that it lets experts in one area encode their knowledge into an easily reusable abstraction, which I think would entail cache TTLs for use cases like the above.

But couldn't this example be implemented by the function author setting the TTL to the underlying IO operation? Be it `Container.withExec`, `http()`, `git()` and so on. That way, the same model would work even for low-level functions - where the callee definitely can't guess of a good TTL (what's the right TTL for `withExec` for example).

I think we still achieve the desired outcome (experts can encode their knowledge in an easily reusable abstraction) but in a more consistent and easy-to-explain way.

**Comment by sipsma** [comment]
> But couldn't this example be implemented by the function author setting the TTL to the underlying IO operation? Be it Container.withExec, http(), git() and so on. That way, the same model would work even for low-level functions - where the callee definitely can't guess of a good TTL (what's the right TTL for withExec for example).

Not if they are doing the IO directly in their code (i.e. using some third party SDK or using some protocol we don't support in the core API).

**Comment by shykes** [comment]
Yeah that makes sense.

**Comment by shykes** [comment]
> Since we don't yet have consensus on how to handle TTLs, all the more reason to cut it from the first phase of implementation :) I will update the proposal to clarify that.

Done ✅

**Comment by sipsma** [comment]
I surprisingly ended up needing to do some anticipatory handling of cases that will become possible once we enable this as part of the secret/socket isolation: https://github.com/dagger/dagger/pull/7804#issuecomment-2216493990
* Just worth a read by whoever ends up implementing this, important context on handling cases around secrets especially

**Comment by pjmagee** [comment]
i've seem very often in various CLI tools a `--no-cache` arg, will this also be something that function developers, or users of the function could use in order to ensure no caching was done?

**Comment by jgsqware** [comment]
Hello, this would in fact be really help, any idea when it can be available?

**Comment by vmaffet-fast** [comment]
:wave: Being able to set caching TTL on all objects would be very nice. For instance, we have a module that generates GitHub App installation tokens, which are valid for 1h. Currently, if your pipeline takes longer than one hour, the token won't be regenerated because the function call was cached. Hence, we have to resort to adding optional cache-busting parameters :(
